File: src/main/java/io/github/cdgeass/formatter/visitor/AbstractCustomVisitor.java
package io.github.cdgeass.formatter.visitor;

import io.github.cdgeass.constants.StringConstants;

/**
 * @author cdgeass
 * @since 2020-06-12
 */
public abstract class AbstractCustomVisitor {

    private final int level;
    private final String tabCharacter;
    private final String preTabCharacter;

    private final StringBuilder sqlStringBuilder;

    protected AbstractCustomVisitor(int level) {
        this.level = level;
        this.tabCharacter = StringConstants.TAB_CHARACTER.repeat(Math.max(0, level));
        this.preTabCharacter = StringConstants.TAB_CHARACTER.repeat(Math.max(0, level - 1));
        this.sqlStringBuilder = new StringBuilder();
    }

    protected AbstractCustomVisitor append(String statement) {
        sqlStringBuilder.append(statement);
        return this;
    }

    protected AbstractCustomVisitor appendTab() {
        sqlStringBuilder.append(tabCharacter);
        return this;
    }

    protected AbstractCustomVisitor appendPreTab() {
        sqlStringBuilder.append(preTabCharacter);
        return this;
    }

    protected int currentLevel() {
        return level;
    }

    protected int nextLevel() {
        return level + 1;
    }

    @Override
    public String toString() {
        return sqlStringBuilder.toString();
    }
}

File: src/main/kotlin/io/github/cdgeass/formatter/visitor/AbstractCustomVisitor.kt
package io.github.cdgeass.formatter.visitor

import io.github.cdgeass.constants.StringConstants

/**
 * jsalparser visitor 抽象
 * 计算换行缩进
 *
 * @author cdgeass
 * @since 2021-03-03
 */
abstract class AbstractCustomVisitor(
    private val level: Int,
    private val tab: String = StringConstants.TAB_CHARACTER.repeat(level),
    private val preTab: String = StringConstants.TAB_CHARACTER.repeat((level - 1).coerceAtLeast(0)),
    private val stringBuilder: StringBuilder = StringBuilder()
) {

    fun append(statement: String): AbstractCustomVisitor {
        stringBuilder.append(statement)
        return this
    }

    fun appendTab(linefeed: Boolean = true): AbstractCustomVisitor {
        if (linefeed && stringBuilder.isNotBlank()) {
            stringBuilder.append(LINE_SEPARATOR)
        }
        if (level != 0) {
            stringBuilder.append(tab)
        }
        return this
    }

    fun appendPreTab(): AbstractCustomVisitor {
        stringBuilder.append(preTab)
        return this
    }

    fun currentLevel(): Int {
        return level
    }

    fun nextLevel(): Int {
        return level + 1
    }

    override fun toString(): String {
        return stringBuilder.toString()
    }

}